### DISCUSSION TITLE
API for Token Session Handler

### DISCUSSION THREAD

**Discussion by wavvves** [woocommerce/woocommerce#44947] [WooCommerce, Blocks, and Site Editing]
We currently have a [custom handler for token-based sessions](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/src/StoreApi/SessionHandler.php), extending WC_Session, but some code is popping out using resources exclusive to the highly cookie-opinionated WC_Session_Handler.

1. We have the https://github.com/woocommerce/woocommerce-blocks/pull/9163 that will keep errors for methods that don't exist on the cookie base session object (eg, trying an invalid method for both classes)
2. On that same PR, if the method exists for cookie but not token class it will suppress the error, sending a doing_it_wrong log entry so a dev knows to take into account the different session handler

Now there are public methods that make sense for both classes (eg, `has_session()` ) and some that do not (eg, `init_session_cookie()`). 

Should we:
1. Port what makes sense and break support for what doesn't, using doing_it_wrong to warn devs to update their code?
2. Wrap everything, even the cookie methods? (It won't be possible to port the same functionality, which may lead to silent bugs/security holes)
3. Work together with core, enforcing proper abstract classes / interfaces we could follow? (may be too late for this now)
4. Any other suggestions? 🤔

Related issues and Slack thread:
https://github.com/woocommerce/woocommerce-blocks/issues/9114
https://github.com/woocommerce/woocommerce-blocks/issues/9116
p1689834226825149-slack-C7U3Y3VMY

**Comment by mikejolley** [comment]
> Work together with core, enforcing proper abstract classes / interfaces we could follow? (may be too late for this now)

@wavvves The abstract `WC_Session` is not opinionated. It seems like cleaning up core to not rely on methods from `WC_Session_Handler` would be best, moving functionality to `WC_Session` that is shared. This shouldn't break backwards compatibility because `WC_Session` is already extended by other session handlers.

**Reply by wavvves** [reply]
>This shouldn't break backwards compatibility because WC_Session is already extended by other session handlers.

That might be tricky to achieve, methods are explicitly named. If we're going to wrap stuff up, it will still be filling holes solution, done in core. If we're filling holes, we might just do it from our side, as it is more manageable and less risky, right?
